pallets/entities/src/lib.rs

```diff
@@ -0,0 +1,164 @@
+	use frame_system::pallet_prelude::*;
+	use fragments_pallet::{Fragment, Fragments, FragmentOwner, DetachedFragments};
+
+	/// Configure the pallet by specifying the parameters and types on which it depends.
+	#[pallet::config]
+	pub trait Config: frame_system::Config + fragments_pallet::Config {
+		/// Because this pallet emits events, it depends on the runtime's definition of an event.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	// fragment-hash to entity-hash-sequence
+	#[pallet::storage]
+	pub type Fragment2Entities<T: Config> = StorageMap<_, Blake2_128Concat, Hash256, Vec<Hash256>>;
+
+	// entity-hash to entity-data
+	#[pallet::storage]
+	pub type Entities<T: Config> = StorageMap<_, Blake2_128Concat, Hash256, EntityData>;
+
+	// entity-hash to entity-id to entity-instance-data
+	#[pallet::storage]
+	pub type EntityInstances<T: Config> = StorageDoubleMap<_, Blake2_128Concat, Hash256, Blake2_128Concat, u128, EntityInstanceData>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Event documentation should end with an array that provides descriptive names for event
+		/// parameters. [something, who]
+		EntityAdded(T::AccountId),
+	}
+
+	// Errors inform users that something went wrong.
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Fragment not found
+		FragmentNotFound,
+		/// Fragment owner not found
+		FragmentOwnerNotFound,
+		/// No Permission
+		NoPermission,
+		/// Fragment is already detached
+		FragmentDetached,
+		/// Entity already exist
+		EntityAlreadyExist
+	}
+
+	// Dispatchable functions allows users to interact with the pallet and invoke state changes.
+	// These functions materialize as "extrinsics", which are often compared to transactions.
+	// Dispatchable functions must be annotated with a weight and must return a DispatchResult.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+
+		#[pallet::weight(<T as Config>::WeightInfo::create())]
+		pub fn create(
+			origin: OriginFor<T>,
+			fragment_hash: Hash256,
+			metadata: EntityMetadata,
+			unique: bool,
+			mutable: bool,
+			max_supply: Option<Compact<u128>>
+		) -> DispatchResult {
+
+			let who = ensure_signed(origin)?;
+			let fragment:Fragment<T::AccountId> = <Fragments<T>>::get(fragment_hash).ok_or(Error::<T>::FragmentNotFound)?;
+
+			let fragment_owner: T::AccountId = match fragment.owner {
+				FragmentOwner::User(owner) => Ok(owner),
+				_ => Err(Error::<T>::FragmentOwnerNotFound),
+			}?;
+
+			ensure!(
+				who == fragment_owner,
+				Error::<T>::NoPermission
+			);
+
+			ensure!(
+				!<DetachedFragments<T>>::contains_key(&fragment_hash),
+				Error::<T>::FragmentDetached
+			);
+
+			let hash = blake2_256(
+				&[
+					&fragment_hash[..],
+					&metadata.name.encode(),
```

Comment:
We need to expand this a bit but for now it's good, maybe add a `//TODO` warning though